main.tf

```diff
@@ -22,6 +22,13 @@ resource "google_pubsub_topic" "topic" {
   project = var.project_id
   name    = var.topic
   labels  = var.topic_labels
+
+  dynamic "message_storage_policy" {
+    for_each = var.message_storage_policy
```

Comment:
I decided to use such structure, because `for_each` [accepts a map or a set](
 https://www.terraform.io/docs/configuration/resources.html#for_each-multiple-resource-instances-defined-by-a-map-or-set-of-strings)